### PR TITLE
Analyzer: fix logical error with set in array join

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 1ed05f1bdd5dec44ac57a0fe88d18ce7cac509b0 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56587/1ed05f1bdd5dec44ac57a0fe88d18ce7cac509b0/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6824148303">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56587/1ed05f1bdd5dec44ac57a0fe88d18ce7cac509b0/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by alexey-milovidov — APPROVED** [review]

File: src/Planner/TableExpressionData.h [hunk 1]
```diff
@@ -270,6 +270,16 @@ class TableExpressionData
         filter_actions = std::move(filter_actions_value);
     }
 
+    const ActionsDAGPtr & getArrayJoinExpressionActions() const
+    {
+        return array_join_actions;
+    }
+
+    void setArrayJoinExpressionActions(ActionsDAGPtr array_join_actions_value)
```
Comment (kitaisreal) [inline src/Planner/TableExpressionData.h:278]: Please move this to `PlannerJoinTree.cpp`.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Planner/TableExpressionData.h (hunk 1 above)
Comment (kitaisreal) [reply src/Planner/TableExpressionData.h:278]: That way fix will work for ARRAY JOIN and JOIN.

File: src/Planner/PlannerJoinTree.h [hunk 2]
```diff
@@ -16,6 +16,7 @@ struct JoinTreeQueryPlan
     QueryPlan query_plan;
     QueryProcessingStage::Enum from_stage;
     std::set<std::string> used_row_policies;
+    std::vector<ActionsDAGPtr> actions_dags = {};
```
Comment (kitaisreal) [inline src/Planner/PlannerJoinTree.h:19]: Vector has default constructor, so we can omit it.

**Review by kitaisreal — CHANGES_REQUESTED** [review]

File: src/Planner/PlannerJoinTree.cpp [hunk 3]
```diff
@@ -1450,6 +1466,10 @@ JoinTreeQueryPlan buildQueryPlanForArrayJoinNode(const QueryTreeNodePtr & array_
     }
 
     array_join_action_dag->projectInput();
+
+    std::vector<ActionsDAGPtr> result_actions_to_execute;
+    result_actions_to_execute.push_back(array_join_action_dag);
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Planner/PlannerJoinTree.cpp]: https://github.com/ClickHouse/ClickHouse/blob/11b8395fe7aa734c9ec8d4ee9f40ede995d230d6/src/Planner/PlannerJoinTree.cpp#L1417-L1423

File: src/Planner/PlannerJoinTree.cpp [hunk 4]
```diff
@@ -1410,7 +1410,23 @@ JoinTreeQueryPlan buildQueryPlanForJoinNode(const QueryTreeNodePtr & join_table_
     for (const auto & right_join_tree_query_plan_row_policy : right_join_tree_query_plan.used_row_policies)
         left_join_tree_query_plan.used_row_policies.insert(right_join_tree_query_plan_row_policy);
 
-    return {std::move(result_plan), QueryProcessingStage::FetchColumns, std::move(left_join_tree_query_plan.used_row_policies)};
+    std::vector<ActionsDAGPtr> result_actions_to_execute;
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Planner/PlannerJoinTree.cpp]: https://github.com/ClickHouse/ClickHouse/blob/11b8395fe7aa734c9ec8d4ee9f40ede995d230d6/src/Planner/PlannerJoinTree.cpp#L1417-L1423

File: src/Planner/PlannerJoinTree.h (hunk 2 above)
Comment (vdimir) [reply src/Planner/PlannerJoinTree.h]: It required to specify member names then.

I replaced

```cpp
return {std::move(query_plan), from_stage, std::move(used_row_policies)};
```

to 
```cpp
return JoinTreeQueryPlan{
        .query_plan = std::move(query_plan),
        .from_stage = from_stage,
        .used_row_policies = std::move(used_row_policies),
    };
```

**Pull request merged**